c_src/optics_htable/htable.c

```diff
@@ -0,0 +1,224 @@
+{
+    uint64_t hash = htable_hash(key);
+    htable_resize(ht, probe_window);
+
+    for (size_t i = 0; i < probe_window; ++i) {
+        struct htable_bucket *bucket = &ht->table[(hash + i) % ht->cap];
+
+        if (!bucket->key) continue;
+        if (bucket->key->n != key->n) continue;
+        if (memcmp(bucket->key->bytes, key->bytes, key->n)) continue;
+
+        return (struct htable_ret) { .ok = true, .value = bucket->value };
+    }
+
+    return (struct htable_ret) { .ok = false };
+}
+
+struct htable_ret htable_put(struct htable *ht, const struct Key *key, uint64_t value)
+{
+    uint64_t hash = htable_hash(key);
+    htable_resize(ht, probe_window);
+
+    struct htable_bucket *empty = NULL;
+
+    for (size_t i = 0; i < probe_window; ++i) {
+        struct htable_bucket *bucket = &ht->table[(hash + i) % ht->cap];
+
+        if (bucket->key) {
+            if (bucket->key->n != key->n) continue;
+            if (memcmp(bucket->key->bytes, key->bytes, key->n)) continue;
+            return (struct htable_ret) { .ok = false, .value = bucket->value };
+        }
+
+        if (!empty) empty = bucket;
+    }
+
+    if (empty) {
+        ht->len++;
+        empty->key = key;
+        empty->value = value;
+        return (struct htable_ret) { .ok = true };
+    }
+
+    htable_resize(ht, ht->cap * 2);
+    return htable_put(ht, key, value);
+}
+
+struct htable_ret htable_del(struct htable *ht, const struct Key *key)
+{
+    uint64_t hash = htable_hash(key);
+    htable_resize(ht, probe_window);
+
+    for (size_t i = 0; i < probe_window; ++i) {
+        struct htable_bucket *bucket = &ht->table[(hash + i) % ht->cap];
+
+        if (!bucket->key) continue;
+        if (bucket->key->n != key->n) continue;
+        if (memcmp(bucket->key->bytes, key->bytes, key->n)) continue;
+
+        ht->len--;
+        free((struct Key *) bucket->key);
+        bucket->key = NULL;
+        return (struct htable_ret) { .ok = true, .value = bucket->value };
+    }
+
+    return (struct htable_ret) { .ok = false };
+}
+
+
+struct htable_bucket * htable_next(
+        struct htable *ht, struct htable_bucket *bucket)
+{
+    if (!ht->table) return NULL;
+
+    size_t i = 0;
+    if (bucket) i = (bucket - ht->table) + 1;
+
+    for (; i < ht->cap; ++i) {
+        bucket = &ht->table[i];
+        if (bucket->key) return bucket;
+    }
+
+    return NULL;
+}
+
+void htable_itr_init(struct htable *ht, struct htable_iterator *itr)
+{
+    itr->bucket = NULL;
+    itr->ht = ht;
+    itr->is_valid = true;
+}
+
+struct htable_ret htable_itr_next(struct htable_iterator *itr)
+{
+    if(itr->is_valid && 
+       (itr->bucket = htable_next(itr->ht, itr->bucket)) != NULL) {
+        return (struct htable_ret) {.ok = true, .value = itr->bucket->value};
+    }
+    itr->is_valid = false;
+    return (struct htable_ret) {.ok = false};
+}
+
+void htable_itr_remove(struct htable_iterator *itr)
+{
+    free((struct Key *) itr->bucket->key);
```

Comment:
You should probably make sure that the key is not null before you try to remove it (the free won't fail but the length decrement would screw things up).